### ISSUE TITLE
elasticsearch-plugin install hang without any reason (no high cpu, nothing)

### ISSUE THREAD

**Issue by PMickael** [elastic/elasticsearch#80480]
**Elasticsearch version** (`bin/elasticsearch --version`): `Version: 7.13.2, Build: default/deb/4d960a0733be83dd2543ca018aa4ddc42e956800/2021-06-10T21:01:55.251515791Z, JVM: 16`

**Plugins installed**: []

**JVM version** (`java -version`): `openjdk 16 2021-03-16
OpenJDK Runtime Environment AdoptOpenJDK (build 16+36)
OpenJDK 64-Bit Server VM AdoptOpenJDK (build 16+36, mixed mode, sharing)`

**OS version** (`uname -a` if on a Unix-like system): `Linux lk-elastic-data-6 4.19.0-18-amd64 #1 SMP Debian 4.19.208-1 (2021-09-29) x86_64 GNU/Linux`

**Description of the problem including expected versus actual behavior**: 

Context, I would like to add 3 nodes to an existing cluster, **all of the 3 servers** are fresh install of debian10, and all of them get the same behavior

**Steps to reproduce**:

 1. Install any plugin
 `/usr/share/elasticsearch/bin/elasticsearch-plugin install analysis-kuromoji`
 2. Hang for a long period without any reason

For repository-gcs it hang for 10-15min before asking y/N
```-> Installing repository-gcs
-> Downloading repository-gcs from elastic
[=================================================] 100%
@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@     WARNING: plugin requires additional permissions     @
@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@
* java.lang.RuntimePermission accessDeclaredMembers
* java.lang.RuntimePermission setFactory
* java.lang.reflect.ReflectPermission suppressAccessChecks
* java.net.SocketPermission * connect,resolve
See http://docs.oracle.com/javase/8/docs/technotes/guides/security/permissions.html
for descriptions of what these permissions allow and the associated risks.

Continue with installation? [y/N]
```

Additionnal information, installing with file:///path/plugin.zip works without hanging, and the plugin install do not respect os proxy I was trying to follow if it was an http request that hang.

**Provide logs (if relevant)**:

2 jstack

https://gist.github.com/PMickael/293f0bc431926fe95f7fe15332d72145
https://gist.github.com/PMickael/0f05418a32b9999f04d274d407af1533

**Comment by DaveCTurner** [comment]
@grcevski this is a recurring problem that users find hard to debug:

- this report
- three users reporting this in https://github.com/elastic/elasticsearch/issues/68419
- https://discuss.elastic.co/t/elasticsearch-plugin-install-plugin-hangs/277732
- perhaps https://discuss.elastic.co/t/elasticsearch-plugin-install-repository-gcs-is-stuck/270537

Although the fundamental problem is environmental, it's far from obvious from the tool behaviour and so it basically looks like a bug. I think we should do more to help users understand why the tool is waiting and guide them towards investigating the problem in their environment. I'm reopening this for now to continue the conversation.

**Comment by PMickael** [comment]
@grcevski, @DaveCTurner 
Like said in my initial post, the problem is highly probable linked to http call error/problem/hanging since it do not respect os proxy, for example a DNS resolve hang, I do not take the time to investigate the install script but take a look around this.

And so the workaround is download the package manually and install it with trick of local url with file:///

**Comment by grcevski** [comment]
Thanks for those pointers, I didn't realize this is more common than a single instance. I'll investigate this in more depth.

**Comment by mmenozzi** [comment]
Like said [here](https://github.com/elastic/elasticsearch/issues/68419#issuecomment-1035997900) it seems that this issue, at least on Ubuntu, is because Elasticsearch waits for a random string from `/dev/random` instead of `/dev/urandom`.

@DaveCTurner do you have any news on that? I experienced the issue again now. Note that this problem makes difficult to roll out Elasticsearch version upgrades because when you upgrade you have to remove and then re-install every plugin. With this issue this takes a lot of time while Elasticsearch is down.

**Comment by DaveCTurner** [comment]
It's fundamentally a system config issue @mmenozzi. On a properly configured system it should be possible to run `dd if=/dev/random of=/dev/null bs=1024 count=1 iflag=fullblock` in minimal time. If this command blocks then that's a problem you need to address. This isn't really the place to debug your system config, but hopefully it helps to note that Linux can gather entropy from numerous sources (e.g. the hardware RNG built into most modern CPUs) if configured properly. Try looking for things like `rng-tools`.

**Comment by mmenozzi** [comment]
@DaveCTurner ok but it's the default config of the most common Linux distribution so many user will experience this problem. Look also what has been written [here](https://superuser.com/a/359601). Don't you think is better to use `/dev/random` instead of `/dev/urandom`?

Even you [here](https://github.com/elastic/elasticsearch/issues/68419#issuecomment-1035967932) said:

> I would expect that it reads from /dev/urandom which blocks less than /dev/random but doesn't guarantee never to block I believe.

**Comment by mmenozzi** [comment]
Anyway if someone here has the same issue I solved it by just installing `haveged` (`apt install haveged`). The plugin installation is much quicker now.

**Comment by grcevski** [comment]
I was able to reproduce the problem with an older Ubuntu kernel (5.4 in my case). The debian 10 uses even older version, so I'm guessing this is going to be a similar issue there. On remote systems there simply isn't enough entropy from mouse or keyboard interaction for the older /dev/random to produce enough data and it simply blocks when we try to get a random number. I confirmed that switching to /dev/urandom resolves the issue, as well as the suggested solution by @mmenozzi to install haveged.

Haveged is already integrated in any kernel including and after 5.6 and those work very fast.

I think we should add a timer around the verify signature call in our InstallPluginCommand and inform the users that this is taking unreasonably long and it could be an insufficient entropy problem with random numbers generation, along with providing guidance on how to resolve the system issue. @DaveCTurner does this work, or you had a different approach in mind?
